source/common/router/config_impl.cc

```diff
@@ -65,6 +67,82 @@ Optional<uint64_t> HashPolicyImpl::generateHash(const Http::HeaderMap& headers)
   return hash;
 }
 
+HeaderFormatterSharedPtr RequestHeaderParser::parseInternal(const std::string& format) {
+	std::string variable_name;
+	if (format.find("%") == 0) {
+		size_t last_occ_pos = format.rfind("%");
+		if (last_occ_pos == std::string::npos || last_occ_pos <= 1) {
+			throw EnvoyException(fmt::format("Incorrect configuration: {}. Expected the variable to be of format %<variable_name>%",format));
+		}
+		variable_name = format.substr(1, last_occ_pos - 1);
+
+
+	}else {
+		HeaderFormatterSharedPtr plain_header_formatter_shard_ptr (new PlainHeaderFormatter(format));
+		return plain_header_formatter_shard_ptr;
+	}
+	HeaderFormatterSharedPtr request_header_formatter_shard_ptr(new RequestHeaderFormatter(variable_name));
+	return request_header_formatter_shard_ptr;
+}
+
+RequestHeaderParserSharedPtr RequestHeaderParser::parse(const Json::Object& config) {
+	RequestHeaderParser* parser = new RequestHeaderParser();
```

Comment:
In this use case you probably want to use std::make_shared<RequestHeaderParser>.  This is the safest way to use shared_ptr, and is exception safe (if an exception is thrown in the code between line 89 and line 100, there will be a leak).